WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainStateHolder.kt

```diff
@@ -16,161 +16,215 @@
-                            showSpeakerNotSupported()
-                            AppState.Ready(transitionInstantly = false)
+                    }
+                AppAction.PlayClicked ->
+                    when (oldState) {
+                        is AppState.Ready,
+                        AppState.PlayingMusic,
+                        AppState.Recording -> {
+                            if (speakerIsSupported()) {
+                                appState = AppState.PlayingVoice
+
+                                soundRecorder.play()
+
+                                appState = AppState.Ready(transitionInstantly = false)
+                            } else {
+                                showSpeakerNotSupported()
+                                appState = AppState.Ready(transitionInstantly = false)
+                            }
                         }
-                    // If we were already playing, transition back to ready
-                    AppState.PlayingMusic -> AppState.Ready(transitionInstantly = false)
-                }
-            AppAction.PlayClicked ->
-                when (oldState) {
-                    is AppState.Ready,
-                    AppState.PlayingMusic,
-                    AppState.Recording -> {
-                        if (speakerIsSupported()) {
-                            AppState.PlayingVoice
-                        } else {
-                            showSpeakerNotSupported()
-                            AppState.Ready(transitionInstantly = false)
+                        // If we were already playing, transition back to ready
+                        AppState.PlayingVoice -> {
+                            appState = AppState.Ready(transitionInstantly = false)
                         }
                     }
-                    // If we were already playing, transition back to ready
-                    AppState.PlayingVoice -> AppState.Ready(transitionInstantly = false)
-                }
-            // We finished doing something, transition back to ready
-            AppAction.PlayMusicFinished,
-            AppAction.PlayRecordingFinished,
-            AppAction.RecordingFinished -> AppState.Ready(transitionInstantly = false)
-            AppAction.PermissionResultReturned ->
-                // Check if the user granted the permission
-                if (
-                    ContextCompat.checkSelfPermission(application, Manifest.permission.RECORD_AUDIO) ==
-                    PackageManager.PERMISSION_GRANTED
-                ) {
-                    // The user granted the permission, continue on to start recording
-                    AppState.Recording
-                } else {
-                    // We have confirmation now that the user denied the permission
-                    isPermissionDeniedMutableStateFlow.value = true
-                    AppState.Ready(transitionInstantly = false)
+                AppAction.PermissionResultReturned ->
```

Comment:
For this version, maybe it's ok to handle permissions in this way. But we should iterate on this implementation and make it more Compose friendly, using the new activityForResult APIs or the Accompanist permission APIs.

Reply:
I am using the `rememberLauncherForActivityResult` API here, and passing it into the state holder. I couldn't use the Accompanist permission API here, since I want to immediately do something upon the user actively accepting the permission, so I want a callback for when the permission result is returned.